liquidity-bot/src/mini_wallet/mod.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) 2023 MobileCoin Inc.
+
+//! A minimalistic "wallet" - a ledger scanner that tracks spendable TxOuts.
+
+mod account_ledger_scanner;
+mod matched_tx_out;
+
+pub use account_ledger_scanner::AccountLedgerScanner;
+pub use matched_tx_out::MatchedTxOut;
+
+use crate::Error;
+use mc_account_keys::AccountKey;
+use mc_blockchain_types::BlockIndex;
+use mc_common::logger::Logger;
+use mc_ledger_db::LedgerDB;
+use mc_transaction_core::ring_signature::KeyImage;
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::HashMap,
+    path::{Path, PathBuf},
+    sync::{Arc, Mutex},
+};
+
+/// An event emitted by the wallet.
+#[derive(Clone, Debug)]
+pub enum WalletEvent {
+    /// A block has been processed.
+    BlockProcessed {
+        /// The index of the block that was processed.
+        block_index: BlockIndex,
+
+        /// Spendable TxOuts found in the block.
+        received_tx_outs: Vec<MatchedTxOut>,
+
+        /// TxOuts that were spent in the block.
+        spent_tx_outs: Vec<MatchedTxOut>,
+    },
+}
+
+/// Callback for notifying the caller of wallet events.
+pub type WalletEventCallback = Arc<dyn Fn(WalletEvent) + Send + Sync>;
+
+/// Internal state of the wallet.
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+pub struct State {
+    next_block_index: BlockIndex,
+    matched_tx_outs: HashMap<KeyImage, MatchedTxOut>,
+}
+impl State {
+    pub fn load(
+        state_file: &PathBuf,
+        default_first_block_index: BlockIndex,
+    ) -> Result<Self, Error> {
+        if state_file.exists() {
+            let bytes = std::fs::read(&state_file)?;
+            Ok(mc_util_serial::deserialize(&bytes)?)
+        } else {
+            Ok(Self {
+                next_block_index: default_first_block_index,
+                ..Default::default()
+            })
+        }
+    }
+    pub fn save(&self, path: &PathBuf) -> Result<(), Error> {
+        let bytes = mc_util_serial::serialize(&self)?;
+        std::fs::write(&path, bytes)?;
+        Ok(())
+    }
+}
+
+pub struct MiniWallet {
+    state: Arc<Mutex<State>>,
+    _account_ledger_scanner: AccountLedgerScanner,
+}
+
+impl MiniWallet {
+    pub fn new(
+        state_file: impl AsRef<Path>,
+        ledger_db: LedgerDB,
+        account_key: AccountKey,
+        default_first_block_index: BlockIndex,
```

Comment:
It might be worth calling out that the default first block index is only used when the state file doesn't load. Though maybe the word default is already doing that.

Reply:
added a comment that documents this